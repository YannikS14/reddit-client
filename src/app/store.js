import { configureStore } from '@reduxjs/toolkit';
import postsReducer from '../features/posts/postsSlice';
import subredditReducer from '../features/subreddits/subredditsSlice';
import searchFieldReducer from '../features/searchField/searchFieldSlice';

export const store = configureStore({
  reducer: {
    posts: postsReducer,
    subreddits: subredditReducer,
    searchField: searchFieldReducer,
  },
});
