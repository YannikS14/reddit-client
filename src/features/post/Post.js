import moment from 'moment';
import React, { useState } from 'react';
import Annotation from '../../assets/Annotation';
import Arrow from '../../assets/Arrow';

export default function Post({ post }) {
  const kFormatter = (num) => {
    return Math.abs(num) > 999
      ? Math.sign(num) * (Math.abs(num) / 1000).toFixed(1) + 'k'
      : Math.sign(num) * Math.abs(num);
  };

  const postDate = moment(post.data.created * 1000).fromNow();
  const upvotes = kFormatter(post.data.ups);
  const commentsAmount = kFormatter(post.data.num_comments);

  const [vote, setVote] = useState('');

  const handleClick = (e) => {
    vote === e.target.id ? setVote('') : setVote(e.target.id);
  };

  return (
    <article className="mb-8 p-6 bg-white rounded-md shadow-lg flex">
      <div id="votes" className="pr-6 flex flex-col items-center">
        <button
          id="upvote"
          onClick={(e) => handleClick(e)}
          className="text-gray-500 hover:text-green-500"
        >
          <Arrow
            className={`transform ${
              vote === 'upvote' ? 'text-green-500 scale-125' : ''
            }`}
          />
        </button>
        <p className="my-1.5 text-lg font-bold text-gray-500">
          {upvotes}
        </p>
        <button
          id="downvote"
          onClick={(e) => handleClick(e)}
          className="text-gray-500 hover:text-red-500"
        >
          <Arrow
            className={`transform rotate-180 ${
              vote === 'downvote' ? 'text-red-500 scale-125' : ''
            }
            `}
          />
        </button>
      </div>

      <div id="post-content" className="flex-1">
        <h3 className="mb-4 text-xl font-semibold text-gray-800">
          {post.data.title}
        </h3>
        {!post.data.is_video && post.data.url && (
          <img src={post.data.url} alt="" className="rounded-md" />
        )}
        {post.data.is_video && (
          <video className="max-h-96" controls>
            <source
              src={post.data.secure_media.reddit_video.fallback_url}
              type="video/mp4"
            />
          </video>
        )}
        {post.data.selftext && (
          <p className="mt-4">{post.data.selftext}</p>
        )}
        <hr className="mt-4" />
        <div
          id="post-footer"
          className="flex justify-between px-8 py-2"
        >
          <div className="font-semibold text-gray-800 text-sm">
            {post.data.author}
          </div>
          <div className="text-gray-800 text-xs">{postDate}</div>
          <div className="flex text-gray-800 text-sm">
            <button
              id="comments"
              className="text-gray-500 hover:text-gray-800"
            >
              <Annotation className="mr-1" />
            </button>
            {commentsAmount}
          </div>
        </div>
      </div>
    </article>
  );
}
