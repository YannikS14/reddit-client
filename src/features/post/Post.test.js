import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../../app/store';
import { render, screen, fireEvent } from '../test-utils';
import Post from './Post';

describe('Post component', () => {
  const testPostMedia = {
    data: {
      created: 1635454,
      ups: 123456,
      num_comments: 123,
      post_hint: 'image',
      url: 'https://linktoimage.test',
      is_video: true,
      media: {
        reddit_video: { fallback_url: 'https://linktovideo.test' },
      },
      selftext: 'Test Text',
      title: 'Test Title',
      author: 'Test author',
      subreddit: 'Testsubreddit',
      id: '123abc',
    },
  };

  it('displays the PostModal component after clicking the button', async () => {
    render(<Post post={testPostMedia} />);

    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    fireEvent.click(screen.getByRole('heading', { level: 2 }));
    expect(await screen.findByRole('dialog')).toBeInTheDocument();
  });

  it('displays the comments after clicking the button', async () => {
    render(<Post post={testPostMedia} />);

    expect(
      screen.queryByTestId('comments-wrapper'),
    ).not.toBeInTheDocument();
    fireEvent.click(screen.getByLabelText('Comments'));
    expect(
      await screen.findByTestId('comments-wrapper'),
    ).toBeInTheDocument();
  });

  it("doesn't render the markdown component, if post text is empty", () => {
    let emptyPostText = testPostMedia;
    emptyPostText.data.selftext = '';

    render(<Post post={emptyPostText} />);
    expect(screen.queryByText('Test Text')).not.toBeInTheDocument();
  });

  it('matches the snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Post post={testPostMedia} />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
