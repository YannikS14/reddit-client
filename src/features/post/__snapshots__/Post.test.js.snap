// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component matches the snapshot 1`] = `
<article
  className="mb-8 p-6 bg-white text-gray-800 dark:bg-gray-700 dark:text-gray-50 rounded-md shadow-lg hover:shadow-xl transition-shadow duration-300 flex z-20 "
>
  <div
    className="pr-6 flex flex-col items-center"
    id="votes"
  >
    <button
      aria-label="Upvote"
      className="text-gray-500 dark:text-gray-300 hover:text-green-500"
      onClick={[Function]}
    >
      <svg
        className="h-6 w-6 pointer-events-none transform transition duration-300 "
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 11l3-3m0 0l3 3m-3-3v8m0-13a9 9 0 110 18 9 9 0 010-18z"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        />
      </svg>
    </button>
    <p
      className="my-1.5 text-lg font-bold"
    >
      123.5k
    </p>
    <button
      aria-label="Downvote"
      className="text-gray-500 dark:text-gray-300 hover:text-red-500"
      onClick={[Function]}
    >
      <svg
        className="h-6 w-6 pointer-events-none transform transition duration-300 rotate-180 
    "
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9 11l3-3m0 0l3 3m-3-3v8m0-13a9 9 0 110 18 9 9 0 010-18z"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
        />
      </svg>
    </button>
  </div>
  <div
    className="flex-1 overflow-hidden"
    id="post-content"
  >
    <h2
      className="mb-4 text-xl font-semibold cursor-pointer"
      onClick={[Function]}
    >
      Test Title
    </h2>
    <img
      alt=""
      className="rounded-md max-h-80"
      src="https://linktoimage.test"
    />
    <video
      className="max-h-80"
      controls={true}
    >
      <source
        src="https://linktovideo.test"
        type="video/mp4"
      />
    </video>
    <hr
      className="mt-4"
    />
    <div
      className="flex flex-wrap justify-between px-8 py-2 gap-2"
      id="post-footer"
    >
      <div
        className="font-semibold text-sm"
      >
        Test author
      </div>
      <div
        className="text-xs"
      >
        52 years ago
      </div>
      <div
        className="flex text-sm"
      >
        <button
          aria-label="Comments"
          className="text-gray-500 dark:text-gray-300 hover:text-gray-800 dark:hover:text-gray-50"
          onClick={[Function]}
        >
          <svg
            className="h-6 w-6 pointer-events-none mr-1 transition duration-300"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 8h10M7 12h4m1 8l-4-4H5a2 2 0 01-2-2V6a2 2 0 012-2h14a2 2 0 012 2v8a2 2 0 01-2 2h-3l-4 4z"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
            />
          </svg>
        </button>
        123
      </div>
    </div>
  </div>
</article>
`;
